stdlib/math.pyi

```diff
@@ -26,7 +26,10 @@ def atanh(__x: _SupportsFloatOrIndex) -> float: ...
 if sys.version_info >= (3, 11):
     def cbrt(__x: _SupportsFloatOrIndex) -> float: ...
 
-def ceil(__x: _SupportsFloatOrIndex) -> int: ...
+class _SupportsCeil(Protocol):
+    def __ceil__(self) -> int: ...
+
+def ceil(__x: _SupportsFloatOrIndex | _SupportsCeil) -> int: ...
```

Comment:
You can implement wacky `__ceil__` methods, and `math.ceil` won't complain:

```pycon
>>> class Foo:
...     def __ceil__(self):
...         return 'foo'
...
>>> f = Foo()
>>> import math
>>> math.ceil(f)
'foo'
```

The docs say that `__ceil__` should generally return an integral, but they don't say that it has to be an `int`. Presumably, it would also be okay to return an instance of `numpy.uint32` or whatever. 

Maybe we should do something like this:

```suggestion
class _SupportsCeil(Protocol[_T_co]):
    def __ceil__(self) -> _T_co: ...

@overload
def ceil(__x: _SupportsCeil[_T]) -> _T: ...
@overload
def ceil(__x: _SupportsFloatOrIndex) -> int: ...
```

Reply:
Well, this is a classic "documentation" vs "runtime" thing.
It is documented that: https://docs.python.org/3/reference/datamodel.html#object.__ceil__

> Called to implement the built-in function [round()](https://docs.python.org/3/library/functions.html#round) and [math](https://docs.python.org/3/library/math.html#module-math) functions [trunc()](https://docs.python.org/3/library/math.html#math.trunc), [floor()](https://docs.python.org/3/library/math.html#math.floor) and [ceil()](https://docs.python.org/3/library/math.html#math.ceil). Unless ndigits is passed to __round__() all these methods should return the value of the object truncated to an [Integral](https://docs.python.org/3/library/numbers.html#numbers.Integral) (typically an [int](https://docs.python.org/3/library/functions.html#int)).

I went with "the docs path". Do you think non-int cases are common enough?
Maybe some middle ground like `TypeVar("_T", bound=int)`?

Reply:
Here are current results:

```
» ag 'def __ceil__'
stdlib/builtins.pyi
285:    def __ceil__(self) -> int: ...
343:        def __ceil__(self) -> int: ...

stdlib/fractions.pyi
130:    def __ceil__(a) -> int: ...

stdlib/_decimal.pyi
118:    def __ceil__(self) -> int: ...

stdlib/numbers.pyi
60:    def __ceil__(self) -> int: ...
```

Reply:
Ok, sounds fair! I hope people won't overuse this great power 🪄